x/snapshot/keeper/keeper.go

```diff
@@ -278,51 +289,39 @@ func (k Keeper) executeSnapshot(ctx sdk.Context, counter int64, keyRequirement t
 		)
 	}
 
-	corruptionThreshold := tss.ComputeAbsCorruptionThreshold(keyRequirement.SafetyThreshold, totalShareCount)
-	if corruptionThreshold < 0 ||
-		corruptionThreshold >= totalShareCount.Int64() {
-		return exported.Snapshot{}, fmt.Errorf("invalid corruption threshold: %d, total share count: %d", corruptionThreshold, totalShareCount.Int64())
+	snapshot := exported.NewSnapshot(
+		participants,
+		ctx.BlockTime(),
+		ctx.BlockHeight(),
+		totalShareCount,
+		counter,
+		keyRequirement.KeyShareDistributionPolicy,
+		tss.ComputeAbsCorruptionThreshold(keyRequirement.SafetyThreshold, totalShareCount),
+	)
+	if err := snapshot.Validate(); err != nil {
+		return exported.Snapshot{}, err
 	}
 
-	snapshot := exported.Snapshot{
-		Validators:                 participants,
-		Timestamp:                  ctx.BlockTime(),
-		Height:                     ctx.BlockHeight(),
-		TotalShareCount:            totalShareCount,
-		Counter:                    counter,
-		KeyShareDistributionPolicy: keyRequirement.KeyShareDistributionPolicy,
-		CorruptionThreshold:        corruptionThreshold,
-	}
-	ctx.KVStore(k.storeKey).Set(counterKey(counter), k.cdc.MustMarshalLengthPrefixed(&snapshot))
+	k.setSnapshot(ctx, snapshot)
 
 	return snapshot, nil
 }
 
-func (k Keeper) setLatestCounter(ctx sdk.Context, counter int64) {
-	bz := make([]byte, 8)
-	binary.LittleEndian.PutUint64(bz, uint64(counter))
-
-	ctx.KVStore(k.storeKey).Set([]byte(lastCounterKey), bz)
-}
-
-func counterKey(counter int64) []byte {
-	return []byte(fmt.Sprintf("%s%d", counterPrefix, counter))
-}
-
 // RegisterProxy registers a proxy address for a given operator, which can broadcast messages in the principal's name
 // The proxy will be marked as active and to be included in the next snapshot by default
 func (k Keeper) RegisterProxy(ctx sdk.Context, operator sdk.ValAddress, proxy sdk.AccAddress) error {
```

Comment:
this should be renamed before we go live to something like `ActivateProxy`

Reply:
Agree, I'll just do the keeper's part for now.